### ISSUE TITLE
bug in `score_sample()`

### ISSUE THREAD

**Issue by LeonEricsson** [xjdr-alt/entropix#85]
the log_prob in`score_sample` is calculated over all token positions in the sequence, rather than just _next_token_. The intended behaviour is not stated but this doesn't seem correct. 

```
# current implementation
def score_sample(sample):
           log_prob = jnp.sum(jax.nn.log_softmax(logits) * jax.nn.one_hot(sample, logits.shape[-1]))
           confidence_score = (
               (1 - metrics["logits_entropy"]) * cfg.ada_score_logits_ent +
               (1 - metrics["attn_entropy"]) * cfg.ada_score_attn_ent +
               (1 - metrics["logits_varentropy"]) * cfg.ada_score_logits_vent +
               (1 - metrics["attn_varentropy"]) * cfg.ada_score_attn_vent +
               metrics["agreement"] * cfg.ada_score_agree +
               metrics["interaction_strength"] * cfg.ada_score_int
           )
           return log_prob + confidence_score
```
I suspect we're trying to retrieve the log_prob of the sampled token. However, `logits` has shape [bsz, context, vocab], which means we're summing over the logits (at position `sample`) across the entire context. I suspect this is what we want to do:
```
def score_sample(sample):
           log_prob = jnp.sum(jax.nn.log_softmax(logits[:, -1]) * jax.nn.one_hot(sample, logits.shape[-1]), axis=-1)
           confidence_score = (
               (1 - metrics["logits_entropy"]) * cfg.ada_score_logits_ent +
               (1 - metrics["attn_entropy"]) * cfg.ada_score_attn_ent +
               (1 - metrics["logits_varentropy"]) * cfg.ada_score_logits_vent +
               (1 - metrics["attn_varentropy"]) * cfg.ada_score_attn_vent +
               metrics["agreement"] * cfg.ada_score_agree +
               metrics["interaction_strength"] * cfg.ada_score_int
           )
           return log_prob + confidence_score
```
and to avoid recomputation:
```
log_probs = jax.nn.log_softmax(logits[:, -1])
confidence_score = (
           (1 - metrics["logits_entropy"]) * cfg.ada_score_logits_ent +
           (1 - metrics["attn_entropy"]) * cfg.ada_score_attn_ent +
           (1 - metrics["logits_varentropy"]) * cfg.ada_score_logits_vent +
           (1 - metrics["attn_varentropy"]) * cfg.ada_score_attn_vent +
                       metrics["agreement"] * cfg.ada_score_agree +
                       metrics["interaction_strength"] * cfg.ada_score_int
                   )
def score_sample(sample):
   log_prob = jnp.sum(log_probs * jax.nn.one_hot(sample, logits.shape[-1]), axis=-1)
   return log_prob + confidence_score
```

**Comment by artus-LYTiQ** [comment]
Also note that you do not need to calculate the `confidence_score` as the actual choice is done by 

`sample_scores = jnp.array([score_sample(sample) for sample in samples])`

followed by 

`
    best_sample_idx = jnp.argmax(sample_scores)
`

where the argmax will pick the same sample irrespective of the static `confidence_score`

**Comment by LeonEricsson** [comment]
> Also note that you do not need to calculate the `confidence_score` as the actual choice is done by
> 
> `sample_scores = jnp.array([score_sample(sample) for sample in samples])`
> 
> followed by
> 
> `best_sample_idx = jnp.argmax(sample_scores)`
> 
> where the argmax will pick the same sample irrespective of the static `confidence_score`

yes this is true.

> Bringing this one step further: even if we sampled the correct log_prob from the selected token, this whole `sampling over multiple potential candidate tokens` code reduces to `gib me the token with the highest (log) prob` aka `argmax sampling`, wouldn't it?
> 
> Edit: ok, due to the stochastic nature of _sample you get some variations in the tokens that you are drawing.

yeah you're still sampling with the applied temperature, then selecting from the samples using log probs argmax
